# AirFly Insights — Week 1: Project Initialization and Dataset Setup

**Dataset:** NYC Flights 2013 (`flights.csv`)

**Goals for Week 1:**
- Define project KPIs and workflow
- Load the dataset using pandas
- Explore schema, types, size, and null values
- Perform sampling and memory optimizations

**Dataset Source:** Based on the classic `nycflights13` dataset — all flights departing from New York City (EWR, LGA, JFK) in 2013.

## Project KPIs & Workflow

| KPI | Description |
|-----|-------------|
| **On-time Performance** | % of flights with `dep_delay ≤ 15 min` |
| **Average Arrival Delay** | Mean `arr_delay` across all non-cancelled flights |
| **Cancellation Rate** | % of flights with null `dep_time` |
| **Busiest Routes** | Top origin-destination pairs by frequency |
| **Carrier Reliability** | Avg delay and cancellation rate per carrier |



## 1. Imports & Setup

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

print('pandas version:', pd.__version__)
print('numpy version:', np.__version__)

pandas version: 3.0.0
numpy version: 2.4.2


## 2. Load Dataset

In [2]:
# Load the raw flights dataset
df = pd.read_csv('../data/raw/flights.csv')

print(f'Dataset loaded successfully!')
print(f'Shape: {df.shape[0]:,} rows × {df.shape[1]} columns')

Dataset loaded successfully!
Shape: 336,776 rows × 21 columns


## 3. Schema Exploration

In [3]:
# Column names
print('=== COLUMNS ===')
print(list(df.columns))

=== COLUMNS ===
['id', 'year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay', 'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight', 'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute', 'time_hour', 'name']


In [4]:
# Data types
print('=== DATA TYPES ===')
df.dtypes

=== DATA TYPES ===


id                  int64
year                int64
month               int64
day                 int64
dep_time          float64
sched_dep_time      int64
dep_delay         float64
arr_time          float64
sched_arr_time      int64
arr_delay         float64
carrier               str
flight              int64
tailnum               str
origin                str
dest                  str
air_time          float64
distance            int64
hour                int64
minute              int64
time_hour             str
name                  str
dtype: object

In [5]:
# Dataset info (non-null counts and dtypes)
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 21 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              336776 non-null  int64  
 1   year            336776 non-null  int64  
 2   month           336776 non-null  int64  
 3   day             336776 non-null  int64  
 4   dep_time        328521 non-null  float64
 5   sched_dep_time  336776 non-null  int64  
 6   dep_delay       328521 non-null  float64
 7   arr_time        328063 non-null  float64
 8   sched_arr_time  336776 non-null  int64  
 9   arr_delay       327346 non-null  float64
 10  carrier         336776 non-null  str    
 11  flight          336776 non-null  int64  
 12  tailnum         334264 non-null  str    
 13  origin          336776 non-null  str    
 14  dest            336776 non-null  str    
 15  air_time        327346 non-null  float64
 16  distance        336776 non-null  int64  
 17  hour            33677

### Column Reference

| Column | Type | Description |
|--------|------|-------------|
| `id` | int | Row identifier |
| `year`, `month`, `day` | int | Flight date |
| `dep_time` | float | Actual departure time (HHMM), null = cancelled |
| `sched_dep_time` | int | Scheduled departure time (HHMM) |
| `dep_delay` | float | Departure delay in minutes (negative = early) |
| `arr_time` | float | Actual arrival time (HHMM) |
| `sched_arr_time` | int | Scheduled arrival time (HHMM) |
| `arr_delay` | float | Arrival delay in minutes |
| `carrier` | str | 2-letter IATA carrier code |
| `flight` | int | Flight number |
| `tailnum` | str | Aircraft tail/registration number |
| `origin` | str | Origin airport (EWR, LGA, JFK) |
| `dest` | str | Destination airport |
| `air_time` | float | Flight duration in minutes |
| `distance` | int | Distance between airports (miles) |
| `hour`, `minute` | int | Scheduled departure time components |
| `time_hour` | str | Timestamp string (e.g. `2013-01-01 05:00:00`) |
| `name` | str | Full airline name |

## 4. Sample Rows

In [6]:
# First 5 rows
print('=== HEAD ===')
df.head()

=== HEAD ===


id  year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0   0  2013      1    1    517.00             515       2.00    830.00   
1   1  2013      1    1    533.00             529       4.00    850.00   
2   2  2013      1    1    542.00             540       2.00    923.00   
3   3  2013      1    1    544.00             545      -1.00   1004.00   
4   4  2013      1    1    554.00             600      -6.00    812.00   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819      11.00      UA    1545  N14228    EWR  IAH    227.00   
1             830      20.00      UA    1714  N24211    LGA  IAH    227.00   
2             850      33.00      AA    1141  N619AA    JFK  MIA    160.00   
3            1022     -18.00      B6     725  N804JB    JFK  BQN    183.00   
4             837     -25.00      DL     461  N668DN    LGA  ATL    116.00   

   distance  hour  minute            time_hour                    name  
0      1400     5      15  2013-01-01 05:00:00   United Air Lines Inc.  
1      1416     5      29  2013-01-01 05:00:00   United Air Lines Inc.  
2      1089     5      40  2013-01-01 05:00:00  American Airlines Inc.  
3      1576     5      45  2013-01-01 05:00:00         JetBlue Airways  
4       762     6       0  2013-01-01 06:00:00    Delta Air Lines Inc.

In [7]:
# Random sample
print('=== RANDOM SAMPLE ===')
df.sample(5, random_state=42)

=== RANDOM SAMPLE ===


id  year  month  day  dep_time  sched_dep_time  dep_delay  \
159280  159280  2013      3   25   1929.00            1905      24.00   
189101  189101  2013      4   26    956.00            1000      -4.00   
212435  212435  2013      5   21   1320.00            1309      11.00   
266804  266804  2013      7   18   1222.00            1230      -8.00   
306581  306581  2013      8   29    540.00             545      -5.00   

        arr_time  sched_arr_time  arr_delay carrier  flight tailnum origin  \
159280   2236.00            2217      19.00      UA    1471  N37298    EWR   
189101   1257.00            1334     -37.00      DL    1765  N717TW    JFK   
212435   1430.00            1414      16.00      EV    4129  N11536    EWR   
266804   1357.00            1419     -22.00      EV    5796  N13958    EWR   
306581    921.00             921       0.00      B6     939  N535JB    JFK   

       dest  air_time  distance  hour  minute            time_hour  \
159280  RSW    169.00      1068    19       5  2013-03-25 19:00:00   
189101  SFO    337.00      2586    10       0  2013-04-26 10:00:00   
212435  DCA     39.00       199    13       9  2013-05-21 13:00:00   
266804  CLT     77.00       529    12      30  2013-07-18 12:00:00   
306581  BQN    198.00      1576     5      45  2013-08-29 05:00:00   

                            name  
159280     United Air Lines Inc.  
189101      Delta Air Lines Inc.  
212435  ExpressJet Airlines Inc.  
266804  ExpressJet Airlines Inc.  
306581           JetBlue Airways

## 5. Descriptive Statistics

In [8]:
# Numerical summary
df.describe().T

count      mean      std     min      25%       50%  \
id             336776.00 168387.50 97219.00    0.00 84193.75 168387.50   
year           336776.00   2013.00     0.00 2013.00  2013.00   2013.00   
month          336776.00      6.55     3.41    1.00     4.00      7.00   
day            336776.00     15.71     8.77    1.00     8.00     16.00   
dep_time       328521.00   1349.11   488.28    1.00   907.00   1401.00   
sched_dep_time 336776.00   1344.25   467.34  106.00   906.00   1359.00   
dep_delay      328521.00     12.64    40.21  -43.00    -5.00     -2.00   
arr_time       328063.00   1502.05   533.26    1.00  1104.00   1535.00   
sched_arr_time 336776.00   1536.38   497.46    1.00  1124.00   1556.00   
arr_delay      327346.00      6.90    44.63  -86.00   -17.00     -5.00   
flight         336776.00   1971.92  1632.47    1.00   553.00   1496.00   
air_time       327346.00    150.69    93.69   20.00    82.00    129.00   
distance       336776.00   1039.91   733.23   17.00   502.00    872.00   
hour           336776.00     13.18     4.66    1.00     9.00     13.00   
minute         336776.00     26.23    19.30    0.00     8.00     29.00   

                     75%       max  
id             252581.25 336775.00  
year             2013.00   2013.00  
month              10.00     12.00  
day                23.00     31.00  
dep_time         1744.00   2400.00  
sched_dep_time   1729.00   2359.00  
dep_delay          11.00   1301.00  
arr_time         1940.00   2400.00  
sched_arr_time   1945.00   2359.00  
arr_delay          14.00   1272.00  
flight           3465.00   8500.00  
air_time          192.00    695.00  
distance         1389.00   4983.00  
hour               17.00     23.00  
minute             44.00     59.00

In [9]:
# Categorical / object column summaries
cat_cols = df.select_dtypes(include='object').columns
print(f'Categorical columns: {list(cat_cols)}')
print()
for col in cat_cols:
    print(f'  {col}: {df[col].nunique()} unique values')
    print(f'    Top 5: {list(df[col].value_counts().head().index)}')
    print()

Categorical columns: ['carrier', 'tailnum', 'origin', 'dest', 'time_hour', 'name']

  carrier: 16 unique values
    Top 5: ['UA', 'B6', 'EV', 'DL', 'AA']

  tailnum: 4043 unique values
    Top 5: ['N725MQ', 'N722MQ', 'N723MQ', 'N711MQ', 'N713MQ']

  origin: 3 unique values
    Top 5: ['EWR', 'JFK', 'LGA']

  dest: 105 unique values
    Top 5: ['ORD', 'ATL', 'LAX', 'BOS', 'MCO']

  time_hour: 6936 unique values
    Top 5: ['2013-09-13 08:00:00', '2013-09-20 08:00:00', '2013-09-09 08:00:00', '2013-09-16 08:00:00', '2013-09-23 08:00:00']

  name: 16 unique values
    Top 5: ['United Air Lines Inc.', 'JetBlue Airways', 'ExpressJet Airlines Inc.', 'Delta Air Lines Inc.', 'American Airlines Inc.']



## 6. Null / Missing Value Analysis

In [10]:
# Null count and percentage per column
null_counts = df.isnull().sum()
null_pct = (null_counts / len(df) * 100).round(2)

null_summary = pd.DataFrame({
    'null_count': null_counts,
    'null_pct': null_pct
}).sort_values('null_count', ascending=False)

print('=== NULL ANALYSIS ===')
null_summary[null_summary['null_count'] > 0]

=== NULL ANALYSIS ===


null_count  null_pct
arr_delay        9430      2.80
air_time         9430      2.80
arr_time         8713      2.59
dep_delay        8255      2.45
dep_time         8255      2.45
tailnum          2512      0.75

In [11]:
# Cancelled flights: rows where dep_time is null
cancelled = df[df['dep_time'].isnull()]
print(f'Cancelled flights: {len(cancelled):,}')
print(f'Cancellation rate: {len(cancelled)/len(df)*100:.2f}%')
print(f'Total flights: {len(df):,}')
print(f'Completed flights: {len(df) - len(cancelled):,}')

Cancelled flights: 8,255
Cancellation rate: 2.45%
Total flights: 336,776
Completed flights: 328,521


**Key observation:** The 5 columns with nulls (`dep_time`, `dep_delay`, `arr_time`, `arr_delay`, `air_time`) are ALL caused by **cancelled flights**. There is no random missingness — these rows can be treated consistently in preprocessing.

## 7. Unique Value Exploration

In [12]:
# Origin airports
print('Origin airports:')
print(df['origin'].value_counts())
print()

# Carriers
print('Carriers (count of flights):')
print(df['carrier'].value_counts())
print()

# Months
print('Flights per month:')
print(df['month'].value_counts().sort_index())
print()

# Destination airports
print(f'Unique destinations: {df["dest"].nunique()}')
print('Top 10 destinations:')
print(df['dest'].value_counts().head(10))

Origin airports:
origin
EWR    120835
JFK    111279
LGA    104662
Name: count, dtype: int64

Carriers (count of flights):
carrier
UA    58665
B6    54635
EV    54173
DL    48110
AA    32729
MQ    26397
US    20536
9E    18460
WN    12275
VX     5162
FL     3260
AS      714
F9      685
YV      601
HA      342
OO       32
Name: count, dtype: int64

Flights per month:
month
1     27004
2     24951
3     28834
4     28330
5     28796
6     28243
7     29425
8     29327
9     27574
10    28889
11    27268
12    28135
Name: count, dtype: int64

Unique destinations: 105
Top 10 destinations:
dest
ORD    17283
ATL    17215
LAX    16174
BOS    15508
MCO    14082
CLT    14064
SFO    13331
FLL    12055
MIA    11728
DCA     9705
Name: count, dtype: int64


In [13]:
# Full airline names
print('Airlines in the dataset:')
airlines = df[['carrier', 'name']].drop_duplicates().sort_values('carrier')
for _, row in airlines.iterrows():
    print(f'  {row["carrier"]} → {row["name"]}')

Airlines in the dataset:
  9E → Endeavor Air Inc.
  AA → American Airlines Inc.
  AS → Alaska Airlines Inc.
  B6 → JetBlue Airways
  DL → Delta Air Lines Inc.
  EV → ExpressJet Airlines Inc.
  F9 → Frontier Airlines Inc.
  FL → AirTran Airways Corporation
  HA → Hawaiian Airlines Inc.
  MQ → Envoy Air
  OO → SkyWest Airlines Inc.
  UA → United Air Lines Inc.
  US → US Airways Inc.
  VX → Virgin America
  WN → Southwest Airlines Co.
  YV → Mesa Airlines Inc.


## 8. Memory Optimization

In [14]:
# Memory usage before optimization
mem_before = df.memory_usage(deep=True).sum() / 1024**2
print(f'Memory before optimization: {mem_before:.2f} MB')

df_opt = df.copy()

# Downcast integer columns
int_cols = df_opt.select_dtypes(include=['int64']).columns
for col in int_cols:
    df_opt[col] = pd.to_numeric(df_opt[col], downcast='integer')

# Downcast float columns
float_cols = df_opt.select_dtypes(include=['float64']).columns
for col in float_cols:
    df_opt[col] = pd.to_numeric(df_opt[col], downcast='float')

# Convert low-cardinality object columns to category
cat_candidates = ['carrier', 'origin', 'dest', 'tailnum', 'name']
for col in cat_candidates:
    if col in df_opt.columns:
        df_opt[col] = df_opt[col].astype('category')

# Memory usage after optimization
mem_after = df_opt.memory_usage(deep=True).sum() / 1024**2
print(f'Memory after optimization:  {mem_after:.2f} MB')
print(f'Reduction: {((mem_before - mem_after) / mem_before * 100):.1f}%')

Memory before optimization: 70.66 MB
Memory after optimization:  22.86 MB
Reduction: 67.6%


In [15]:
# Per-column memory comparison
mem_compare = pd.DataFrame({
    'before (KB)': df.memory_usage(deep=True) / 1024,
    'after (KB)': df_opt.memory_usage(deep=True) / 1024
}).round(2)
mem_compare['saved (KB)'] = (mem_compare['before (KB)'] - mem_compare['after (KB)']).round(2)
print(mem_compare)

                before (KB)  after (KB)  saved (KB)
Index                  0.13        0.13        0.00
id                  2631.06     1315.53     1315.53
year                2631.06      657.77     1973.29
month               2631.06      328.88     2302.18
day                 2631.06      328.88     2302.18
dep_time            2631.06     1315.53     1315.53
sched_dep_time      2631.06      657.77     1973.29
dep_delay           2631.06     1315.53     1315.53
arr_time            2631.06     1315.53     1315.53
sched_arr_time      2631.06      657.77     1973.29
arr_delay           2631.06     1315.53     1315.53
carrier             3288.83      329.04     2959.79
flight              2631.06      657.77     1973.29
tailnum             4629.19      713.52     3915.67
origin              3617.71      328.92     3288.79
dest                3617.71      330.02     3287.69
air_time            2631.06     1315.53     1315.53
distance            2631.06      657.77     1973.29
hour        

## 9. Week 1 Summary & Key Insights

| Metric | Value |
|--------|-------|
| Total rows | 336,776 |
| Total columns | 21 |
| Date range | Jan 2013 – Dec 2013 |
| Origin airports | 3 (EWR, LGA, JFK) |
| Airlines | 16 |
| Destination airports | 105 |
| Cancelled flights | ~8,255 (~2.45%) |
| Null columns | 5 (all from cancellations) |

**Key Observations:**
1. **Nulls are structured** — only from cancelled flights; no random missing data
2. **Three NYC origin airports** — EWR (Newark), LGA (LaGuardia), JFK (JFK)
3. **16 carriers** — UA (United), B6 (JetBlue), EV (ExpressJet), DL (Delta) are the busiest
4. **Delays can be negative** — negative `dep_delay` means departed early
5. **Memory can be reduced ~60%** via downcasting and categorical types
